### PR TITLE
Restore sub-sub-apps when needed, even if the sub-app does not need restored

### LINKED ISSUE #15837: Sub-sub-apps may need restored even when the sub-app does not
## Bug Description

Currently, a sub-app is restored if `needRestoration()` is true.
If it is false however, the ability for the sub-sub-apps to be restored is removed.

In particular, if the sub-sub-app is using sub-cycling (but the sub-app is not), then the sub-app will properly have its time reset to the beginning of the time step when the driving app starts the second iteration. On the other hand, the sub-sub-app remains at the end of the time step and no longer solves for the rest of the Picard iterations.

## Steps to Reproduce

Actually, the test `test/tests/multiapps/picard_multilevel/test` already has that faulty behavior. It is subtle which is why it wasn't originally caught.

Here is the (abridged) screen output with `!!!!!!!!` added to show where the problem is: 

```
Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+
| time           | picard_its     |
+----------------+----------------+
|   0.000000e+00 |   1.000000e+00 |
+----------------+----------------+


Time Step 1, time = 1, dt = 1

Backing Up MultiApps on TIMESTEP_END
Beginning backing up MultiApp sub1
sub10: Beginning backing up MultiApp sub2
sub10: Finished backing up MultiApp sub2
Finished backing up MultiApp sub1
Finished Backing Up MultiApps on TIMESTEP_END

Initial Picard Norm: 3.316625e+00

Beginning Picard Iteration 0

Master solve:
 0 Nonlinear |R| = 3.030837e-01
 1 Nonlinear |R| = 9.780799e-07
 2 Nonlinear |R| = 6.309854e-12
 Solve Converged!

Starting Transfers on TIMESTEP_END To MultiApps
Beginning NearestNodeTransfer v1
Finished NearestNodeTransfer v1
Transfers on TIMESTEP_END Are Finished


Executing MultiApps on TIMESTEP_END
Solving MultiApp 'sub1' with target time 1.000000e+00 and dt 1.000000e+00 with auto-advance off
sub10: 
sub10: Time Step 1, time = 1, dt = 1
sub10: 
sub10: Backing Up MultiApps on TIMESTEP_END
sub10: Beginning backing up MultiApp sub2
sub10: Finished backing up MultiApp sub2
sub10: Finished Backing Up MultiApps on TIMESTEP_END
sub10: 
sub10: Initial Picard Norm: 3.316787e+00
sub10: 
sub10: Beginning Picard Iteration 0
sub10: 
sub10: Master solve:
sub10:  0 Nonlinear |R| = 3.078092e+00
sub10:  1 Nonlinear |R| = 6.126159e-07
sub10:  2 Nonlinear |R| = 2.006743e-13
sub10:  Solve Converged!
sub10: 
sub10: No Transfers on TIMESTEP_END To MultiApps
sub10: 
sub10: 
sub10: Executing MultiApps on TIMESTEP_END
sub10: Solving MultiApp 'sub2' with target time 1.000000e+00 and dt 1.000000e+00 with auto-advance off
sub10_sub20: 
sub10_sub20: Time Step 1, time = 0.5, dt = 0.5
sub10_sub20:  0 Nonlinear |R| = 3.071933e+00
sub10_sub20:  1 Nonlinear |R| = 5.167684e-07
sub10_sub20:  2 Nonlinear |R| = 1.466092e-13
sub10_sub20:  Solve Converged!
sub10_sub20: 
sub10_sub20: Time Step 2, time = 1, dt = 0.5
sub10_sub20:  0 Nonlinear |R| = 9.111232e-02
sub10_sub20:  1 Nonlinear |R| = 2.406539e-08
sub10_sub20:  2 Nonlinear |R| = 1.025853e-14
sub10_sub20:  Solve Converged!
sub10: Successfully Solved MultiApp sub2.
sub10: Finished Executing MultiApps on TIMESTEP_END
sub10: 
sub10: 
sub10: Starting Transfers on TIMESTEP_END From MultiApps
sub10: Beginning NearestNodeTransfer v2
sub10: Finished NearestNodeTransfer v2
sub10: Transfers on TIMESTEP_END Are Finished
sub10: 
sub10: Picard Norm after TIMESTEP_END MultiApps: 5.099786e-02
sub10: 
sub10:  0 Picard |R| = 3.316787e+00
sub10:  1 Picard |R| = 5.099786e-02
sub10: 
sub10: Restoring MultiApps on TIMESTEP_END
sub10: Begining restoring MultiApp sub2
sub10: Finished restoring MultiApp sub2
sub10: Finished Restoring MultiApps on TIMESTEP_END
sub10: 
sub10: Beginning Picard Iteration 1
sub10: 
sub10: Master solve:
sub10:  0 Nonlinear |R| = 5.099786e-02
sub10:  1 Nonlinear |R| = 1.751696e-08
sub10:  2 Nonlinear |R| = 8.960277e-15
sub10:  Solve Converged!
sub10: 
sub10: No Transfers on TIMESTEP_END To MultiApps
sub10: 
sub10: 
sub10: Executing MultiApps on TIMESTEP_END
sub10: Solving MultiApp 'sub2' with target time 1.000000e+00 and dt 1.000000e+00 with auto-advance off
sub10_sub20: 
sub10_sub20: Time Step 2, time = 0.5, dt = 0.5
sub10_sub20:  0 Nonlinear |R| = 3.071933e+00
sub10_sub20:  1 Nonlinear |R| = 5.167684e-07
sub10_sub20:  2 Nonlinear |R| = 1.466092e-13
sub10_sub20:  Solve Converged!
sub10_sub20: 
sub10_sub20: Time Step 3, time = 1, dt = 0.5
sub10_sub20:  0 Nonlinear |R| = 9.111232e-02
sub10_sub20:  1 Nonlinear |R| = 2.406539e-08
sub10_sub20:  2 Nonlinear |R| = 1.025853e-14
sub10_sub20:  Solve Converged!
sub10: Successfully Solved MultiApp sub2.
sub10: Finished Executing MultiApps on TIMESTEP_END
sub10: 
sub10: 
sub10: Starting Transfers on TIMESTEP_END From MultiApps
sub10: Beginning NearestNodeTransfer v2
sub10: Finished NearestNodeTransfer v2
sub10: Transfers on TIMESTEP_END Are Finished
sub10: 
sub10: Picard Norm after TIMESTEP_END MultiApps: 8.960277e-15
sub10: 
sub10:  0 Picard |R| = 3.316787e+00
sub10:  1 Picard |R| = 5.099786e-02
sub10:  2 Picard |R| = 8.960277e-15
sub10: Picard converged reason: CONVERGED_ABS
Successfully Solved MultiApp sub1.
Finished Executing MultiApps on TIMESTEP_END


Starting Transfers on TIMESTEP_END From MultiApps
Beginning NearestNodeTransfer v
Finished NearestNodeTransfer v
Transfers on TIMESTEP_END Are Finished

Picard Norm after TIMESTEP_END MultiApps: 5.347295e-02

 0 Picard |R| = 3.316625e+00
 1 Picard |R| = 5.347295e-02

Restoring MultiApps on TIMESTEP_END
Finished Restoring MultiApps on TIMESTEP_END

Beginning Picard Iteration 1

Master solve:
 0 Nonlinear |R| = 5.347295e-02
 1 Nonlinear |R| = 4.491350e-07
 2 Nonlinear |R| = 1.280984e-12
 Solve Converged!

Starting Transfers on TIMESTEP_END To MultiApps
Beginning NearestNodeTransfer v1
Finished NearestNodeTransfer v1
Transfers on TIMESTEP_END Are Finished


Executing MultiApps on TIMESTEP_END
Solving MultiApp 'sub1' with target time 1.000000e+00 and dt 1.000000e+00 with auto-advance off
sub10: 
sub10: Time Step 1, time = 1, dt = 1
sub10: 
sub10: Backing Up MultiApps on TIMESTEP_END
sub10: Beginning backing up MultiApp sub2
sub10: Finished backing up MultiApp sub2
sub10: Finished Backing Up MultiApps on TIMESTEP_END
sub10: 
sub10: Initial Picard Norm: 2.315688e-02
sub10: 
sub10: Beginning Picard Iteration 0
sub10: 
sub10: Master solve:
sub10:  0 Nonlinear |R| = 2.315688e-02
sub10:  1 Nonlinear |R| = 1.432980e-08
sub10:  2 Nonlinear |R| = 3.920191e-15
sub10:  Solve Converged!
sub10: 
sub10: No Transfers on TIMESTEP_END To MultiApps
sub10: 
sub10: 
sub10: Executing MultiApps on TIMESTEP_END
sub10: Solving MultiApp 'sub2' with target time 1.000000e+00 and dt 1.000000e+00 with auto-advance off
sub10: Successfully Solved MultiApp sub2.
!!!!!!!!!!!!!!!!!!!!!
SUB2 DID NOT SOLVE BECAUSE IT WASN'T RESTORED AND THINKS IT'S ALREADY AT THE TARGET TIME
!!!!!!!!!!!!!!!!!!!!!
sub10: Finished Executing MultiApps on TIMESTEP_END
sub10: 
sub10: 
sub10: Starting Transfers on TIMESTEP_END From MultiApps
sub10: Beginning NearestNodeTransfer v2
sub10: Finished NearestNodeTransfer v2
sub10: Transfers on TIMESTEP_END Are Finished
sub10: 
sub10: Picard Norm after TIMESTEP_END MultiApps: 3.920191e-15
sub10: 
sub10:  0 Picard |R| = 2.315688e-02
sub10:  1 Picard |R| = 3.920191e-15
sub10: Picard converged reason: CONVERGED_ABS
Successfully Solved MultiApp sub1.
Finished Executing MultiApps on TIMESTEP_END


Starting Transfers on TIMESTEP_END From MultiApps
Beginning NearestNodeTransfer v
Finished NearestNodeTransfer v
Transfers on TIMESTEP_END Are Finished

Picard Norm after TIMESTEP_END MultiApps: 2.079624e-03

 0 Picard |R| = 3.316625e+00
 1 Picard |R| = 5.347295e-02
 2 Picard |R| = 2.079624e-03
```

## Impact

This means that the converged solution at the end of a given time step may not be the desired converged solution as the sub-sub-app may not have gotten the most up-to-date data transfer.

### PR DESCRIPTION
**Pull request by vincentlaboure**
This PR addresses and closes #15837, as discussed in the issue.

Tag @lindsayad @fdkong

### CODE REVIEW

**Comment by vincentlaboure** [comment]
Should be ready to be reviewed. Creating a test wasn't easy because the final solution is not necessarily changed by the fix.

Thanks @lindsayad for self-assigning!

File: framework/src/multiapps/MultiApp.C [hunk 1]
```diff
@@ -482,6 +482,20 @@ MultiApp::restore()
   }
 }
 
+void
+MultiApp::restoreSubApps()
+{
+  for (unsigned int i = 0; i < _apps.size(); i++)
+  {
+    for (auto & sub_app :
+         _apps[i]->getExecutioner()->feProblem().getMultiAppWarehouse().getObjects())
```
Comment (lindsayad) [inline framework/src/multiapps/MultiApp.C:491]: It's hilarious to me that the `MultiApp` needs to do this in order to get the sub-apps. This is beyond the scope of your PR though

File: framework/src/multiapps/MultiApp.C [hunk 2]
```diff
@@ -482,6 +482,20 @@ MultiApp::restore()
...
+      if (sub_app->needsRestoration())
+        sub_app->restore();
+      else
+        sub_app->restoreSubApps();
```
Comment (lindsayad) [inline framework/src/multiapps/MultiApp.C:495]: I love the recursion!

File: framework/src/multiapps/MultiApp.C [hunk 3]
```diff
@@ -482,6 +482,20 @@ MultiApp::restore()
...
 
+void
+MultiApp::restoreSubApps()
+{
+  for (unsigned int i = 0; i < _apps.size(); i++)
+  {
+    for (auto & sub_app :
+         _apps[i]->getExecutioner()->feProblem().getMultiAppWarehouse().getObjects())
+      if (sub_app->needsRestoration())
+        sub_app->restore();
```
Comment (lindsayad) [inline framework/src/multiapps/MultiApp.C:493]: So if you call `restore` on a sub-app, then that restores all sub-sub-apps as well?

Comment (vincentlaboure) [reply framework/src/multiapps/MultiApp.C:493]: That's my understanding at least ;) 
It seems to me that any app restores everything under it if it is restored.

**Review by lindsayad — COMMENTED** [review]
Here would be what I would do:
1. Add `bool force = false` parameter to `MultiApp::restore` (e.g. it defaults to `false` if a user doesn't provide an argument)
2. `FEProblemBase::restoreMultiApps` becomes:
```c++
for (const auto & multi_app : multi_apps)
  multi_app->restore(force);
```
3. `MultiApp::restore` becomes:
```c++
if (force || needsRestoration())
{
  all_the_existing_restore_code;
}
else
{
  for (unsigned int i = 0; i < _apps.size(); i++)
  {
    for (auto & sub_app :
            _apps[i]->getExecutioner()->feProblem().getMultiAppWarehouse().getObjects())
      sub_app->restore();
  }
}
```
Then you wouldn't need to add your new API. What do you think?

**Comment by vincentlaboure** [comment]
Sounds good to me either way. The only thing is that the default value for `force` should probably be `true` for backward compatibility.
Right now, calling `restore()` meant `restore`; it would become more like `restore unless you shouldn't`... so hopefully it's not too confusing.

**Comment by lindsayad** [comment]
> Sounds good to me either way. The only thing is that the default value for `force` should probably be `true` for backward compatibility.
> Right now, calling `restore()` means `restore`...; it would become more like `restore unless you shouldn't`... so hopefully it's not too confusing.

Ok, let's try making the change where we don't have to introduce the new API. If it becomes too much work, we'll go with what you have here.

**Comment by vincentlaboure** [comment]
Comments should be addressed

**Review by lindsayad — APPROVED** [review]
Nice work! Thanks @vincentlaboure

**Comment by vincentlaboure** [comment]
Thanks for the quick review @lindsayad !

**Pull request merged**